### PR TITLE
BO Migration Contact > Stores - Add command, command handler and behat tests for store adding

### PR DESCRIPTION
**Pull request by tom-combet**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | BO Migration Contact > Stores - Add command, command handler and behat tests for store adding
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | N/A
| How to test?      | Run behat test via : ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s store
| Possible impacts? | N/A

The multilang fields types changes for store ObjectModel is probably a BC break. Can you confirm?

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/28033)

### CODE REVIEW

**Comment by tom-combet** [comment]
I'm really not sure about ValueObject class. It's however the most optimized way I found to do it.

Is it a problem to create a common ValueObject class?
If so, I can move the validate method in PrestaShop Validator class for example. I see one drawback: if value object need Symfony Validator instance to perform more checks in other functions, it will be recreated.

Also, behat tests are not working, I need your help on this! CharacterCleaner is not injected when building TypedRegexValidator.

File: src/Adapter/Store/CommandHandler/AddStoreHandler.php
Comment (Progi1984) [inline src/Adapter/Store/CommandHandler/AddStoreHandler.php]: Don't forget header licences

File: src/Adapter/Store/CommandHandler/AddStoreHandler.php
Comment (Progi1984) [inline src/Adapter/Store/CommandHandler/AddStoreHandler.php]:
```suggestion
class AddStoreHandler extends AbstractObjectModelHandler implements AddStoreHandlerInterface
```

File: src/Core/Domain/Store/Command/AddStoreCommand.php
Comment (Progi1984) [inline src/Core/Domain/Store/Command/AddStoreCommand.php:127]: Please update PHPDoc

File: src/Core/Domain/Store/QueryResult/StoreForEditing.php
Comment (Progi1984) [inline src/Core/Domain/Store/QueryResult/StoreForEditing.php:223]: Don't forget return type hint.

File: src/Core/Domain/Store/ValueObject/Coordinate.php
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/Coordinate.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\Store\ValueObject;
```

File: src/Core/Domain/Store/ValueObject/Coordinate.php
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/Coordinate.php]:
```suggestion
    public function getValue(): float
```

File: src/Core/Domain/Store/ValueObject/Hours.php
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/Hours.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\Store\ValueObject;
```

File: src/Core/Domain/Store/ValueObject/Hours.php
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/Hours.php]:
```suggestion
    public function getValue(): string
```

File: src/Core/Domain/Store/ValueObject/Hours.php
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/Hours.php]:
```suggestion
    public function __construct(string $hours)
```

File: src/Core/Domain/Store/ValueObject/Coordinate.php
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/Coordinate.php]:
```suggestion
    public function __construct(float $coordinate)
```

File: src/Core/Domain/Store/ValueObject/Coordinate.php
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/Coordinate.php:58]:
```suggestion
        $this->coordinate = $coordinate;
```

File: src/Core/Domain/ValueObject/Address.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/Address.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\ValueObject;
```

File: src/Core/Domain/ValueObject/CityName.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/CityName.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\ValueObject;
```

File: src/Core/Domain/ValueObject/GenericName.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/GenericName.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\ValueObject;
```

File: src/Core/Domain/ValueObject/Note.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/Note.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\ValueObject;
```

File: src/Core/Domain/ValueObject/PhoneNumber.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/PhoneNumber.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\ValueObject;
```

File: src/Core/Domain/ValueObject/Postcode.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/Postcode.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\ValueObject;
```

File: src/Core/Domain/ValueObject/Postcode.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/Postcode.php]:
```suggestion
    public function getValue(): string
```

File: src/Core/Domain/ValueObject/ValueObject.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/ValueObject.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\ValueObject;
```

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]:
```suggestion
    public function addNewStoreWithTheFollowingDetails(TableNode $table, string $reference): void
```

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]:
```suggestion
    public function storeShouldHaveTheFollowingDetails(string $reference, TableNode $table): void
```

File: src/Core/Domain/ValueObject/ValueObject.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/ValueObject.php]:
```suggestion
    protected function validate($value, array $constraints, ?string $errorMessage = null): void
```

File: src/Core/Domain/ValueObject/Postcode.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/Postcode.php]:
```suggestion
            new TypedRegex(['type' => TypedRegex::TYPE_POST_CODE]),
```

File: src/Core/Domain/ValueObject/PhoneNumber.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/PhoneNumber.php]:
```suggestion
            new TypedRegex(['type' => TypedRegex::TYPE_PHONE_NUMBER]),
```

File: src/Core/Domain/ValueObject/GenericName.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/GenericName.php]:
```suggestion
            new TypedRegex(['type' => TypedRegex::TYPE_GENERIC_NAME]),
```

File: src/Core/Domain/ValueObject/CityName.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/CityName.php]:
```suggestion
            new TypedRegex(['type' => TypedRegex::TYPE_CITY_NAME]),
```

File: src/Core/Domain/ValueObject/Address.php
Comment (Progi1984) [inline src/Core/Domain/ValueObject/Address.php]:
```suggestion
            new TypedRegex(['type' => TypedRegex::TYPE_ADDRESS]),
```

File: src/Core/Domain/Store/ValueObject/Coordinate.php
Comment (Progi1984) [inline src/Core/Domain/Store/ValueObject/Coordinate.php]:
```suggestion
            new TypedRegex(['type' => 'coordinate']),
```

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@tom-combet Thank you for this contribution. Some feedbacks (and some of them for fixing your behat errors).

File: src/Core/Domain/Store/Command/AddStoreCommand.php
Comment (tom-combet) [reply src/Core/Domain/Store/Command/AddStoreCommand.php:127]: Are you talking about constructor? PHPDoc on top of typehints, isn't it overkill?

File: src/Core/Domain/Store/ValueObject/Coordinate.php
Comment (tom-combet) [reply src/Core/Domain/Store/ValueObject/Coordinate.php:58]: This cast is required since `$coordinate` parameter is a string.

File: src/Core/Domain/Store/Command/AddStoreCommand.php
Comment (Progi1984) [reply src/Core/Domain/Store/Command/AddStoreCommand.php:127]: Yes and No (because old IDE doesn't detect the typehint).

File: src/Core/Domain/Store/QueryResult/StoreForEditing.php
Comment (Progi1984) [inline src/Core/Domain/Store/QueryResult/StoreForEditing.php]:
```suggestion
        array $localizedNames = [],
        array $localizedAddress1 = [],
        array $localizedAddress2 = [],
        array $localizedHours = [],
        array $localizedNotes = []
```

File: src/Core/Domain/Store/ValueObject/Coordinate.php
Comment (Progi1984) [reply src/Core/Domain/Store/ValueObject/Coordinate.php]: @tom-combet Could you add these modifications ?

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks remaining.

File: src/Core/Domain/Store/ValueObject/Coordinate.php
Comment (tom-combet) [reply src/Core/Domain/Store/ValueObject/Coordinate.php]: Are you sure that this is necessary? Type is the default option so it can be passed directly without specifying it in an array.
So it should be functional like this, but maybe it is for clarity? And if so, I'll make the changes of course.

File: src/Core/Domain/Store/Command/AddStoreCommand.php
Comment (tom-combet) [reply src/Core/Domain/Store/Command/AddStoreCommand.php:127]: Ok so I do it everywhere it's "missing"? Sorry to bother, just looks really overkill to me, but I understand the IDE issue.

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (atomiix) [inline tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]:
```suggestion
 * @author    PrestaShop SA and Contributors <contact@prestashop.com>
```

File: src/Core/Domain/ValueObject/ValueObject.php
Comment (atomiix) [inline src/Core/Domain/ValueObject/ValueObject.php]:
```suggestion
abstract class ValueObject
```

**Comment by LouisAUTHIE** [comment]
Hi guys, any news about it ? We need it to continue the migration process

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by Progi1984** [comment]
Ping @PrestaShop/prestashop-maintainers & @PrestaShop/committers for a second review.

File: src/Core/Domain/Store/Command/AddStoreCommand.php
Comment (zuk3975) [inline src/Core/Domain/Store/Command/AddStoreCommand.php]: $shopAssociation missing type

File: src/Adapter/Store/CommandHandler/AbstractStoreHandler.php
Comment (zuk3975) [inline src/Adapter/Store/CommandHandler/AbstractStoreHandler.php:37]: I wouldn't say its worth introducing abstract class just to reuse the constructor if it has no shared methods. 
Though i don't know, maybe its just me. Don't rush changing it before we get more opinions :man_shrugging:

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: shop or store?
I see this different naming, is it the same or different thing?

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: maybe the getShopIdsByReferences() method should be instead of this loop?

**Comment by zuk3975** [comment]
> The multilang fields types changes for store ObjectModel is probably a BC break. Can you confirm?

I think you only added missing type hints, but those fields were always multilang (definition hasn't changed) no? so I'ts not a BC

File: src/Adapter/Store/CommandHandler/AbstractStoreHandler.php
Comment (tom-combet) [reply src/Adapter/Store/CommandHandler/AbstractStoreHandler.php:37]: Ah, I don't agree here: all handlers use the StoreRepository instance. So if we don't use an abstract class, we'll have to copy/paste the argument in the services definitions... not good for me.

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (tom-combet) [reply tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: Different things, we cast here the shops array string (i.e `'[shop1, shop2]'`) into a real array.

File: tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php
Comment (tom-combet) [reply tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: Indeed, I refactored the ids retrieval. Please tell me if it looks good for you now.

Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/Domain/StoreFeatureContext.php]: looks good :+1:

File: src/Core/Domain/Store/Command/AddStoreCommand.php
Comment (zuk3975) [reply src/Core/Domain/Store/Command/AddStoreCommand.php:127]: I agree its probably overkill when all types are clearly defined, **but not with arrays**. You don't see what is the type of array elements, so it is definitely useful to see psalm annotations in such cases (also the phpstan complains when those types mismatch :v:  )

File: src/Core/ConstraintValidator/Factory/ConstraintValidatorFactory.php
Comment (zuk3975) [inline src/Core/ConstraintValidator/Factory/ConstraintValidatorFactory.php:48]:
```suggestion
    public function getInstance(Constraint $constraint): ConstraintValidatorInterface
```
?

**Comment by kpodemski** [comment]
Since we have had no news from you for more than 20 days, I'm closing this PR. Feel free to re-open it if you find the motivation to finish it. Don't hesitate to ping me if you have any questions or doubts.

**Pull request closed without merging**
